conans/client/tools/win.py

```diff
@@ -14,10 +14,93 @@
 from conans.util.env_reader import get_env
 from conans.util.files import decode_text, save, mkdir_tmp
 from conans.unicode import get_cwd
+from conans.model.version import Version
 
 _global_output = None
 
 
+def _visual_compiler_cygwin(output, version):
+    if os.path.isfile("/proc/registry/HKEY_LOCAL_MACHINE/SOFTWARE/Microsoft/Windows/CurrentVersion/ProgramFilesDir (x86)"):
+        is_64bits = True
+    else:
+        is_64bits = False
+
+    if is_64bits:
+        key_name = r'HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\VisualStudio\SxS\VC7'
+    else:
+        key_name = r'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\VisualStudio\SxS\VC7'
+
+    if not os.path.isfile("/proc/registry/" + key_name.replace('\\', '/') + "/" + version):
+        return None
+
+    installed_version = Version(version).major(fill=False)
+    compiler = "Visual Studio"
+    output.success("CYGWIN: Found %s %s" % (compiler, installed_version))
+    return compiler, installed_version
+
+
+@contextmanager
+def _system_registry_key(key, subkey):
+    from six.moves import winreg  # @UnresolvedImport
+    hkey = winreg.OpenKey(key, subkey)
+    try:
+        yield hkey
+    finally:
+        winreg.CloseKey(hkey)
+
+
+def _visual_compiler(output, version):
+    """"version have to be 8.0, or 9.0 or... anything .0"""
+    if platform.system().startswith("CYGWIN"):
+        return _visual_compiler_cygwin(output, version)
+
+    if version == "15":
+        vs_path = os.getenv('vs150comntools')
+        path = vs_path or vs_installation_path("15")
+        if path:
+            compiler = "Visual Studio"
+            output.success("Found %s %s" % (compiler, "15"))
+            return compiler, "15"
+        return None
+
+    version = "%s.0" % version
+    from six.moves import winreg  # @UnresolvedImport
+    try:
+        try:
+            with _system_registry_key(winreg.HKEY_LOCAL_MACHINE,
```

Comment:
I think this context manager adds many complexity to the function with many try/catch blocks. I'd rather have a normal function `_system_registry_key(key, subkey, query)` that will make the checks inside try/catch and return the query value or `None` in case it fails. This will cover the two use cases of getting the `installation_folder` and the `is_64bits `

Reply:
changed `_system_registry_key` to function, this simplifies things a lot!